Handle web socket notifications on patient dashboard & archive

nmajor25 opened the pull request:
Shortcut Story ID: [sc-59837]

- [x] Handle ws notifications on patient dashboard
- [x] Handle ws notifications on patient archive
- [x] Use `NIL_UUID` for action.flow filter
- [x] Manually test functionality/edge cases in local dev environment
- [x] Add dashboard cypress tests
- [x] Add archive cypress tests
- [x] Fix failing tests - web socket messages intermittently not coming through at beginning of test
- [x] Rebase branch & use new service request implemented in https://github.com/RoundingWell/care-ops-frontend/pull/1417

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

## Summary by CodeRabbit

- **New Features**
	- Enabled real-time updates for patient data on the archive and dashboard, ensuring that changes to patient flows and actions are reflected live.
	- Improved the visual handling of completed items with smooth fade-out effects for a better user experience.
- **Tests**
	- Expanded test coverage to validate that real-time notifications update the UI consistently for both flows and actions.
	- Added new test specifications for socket notifications related to flows and actions in both the patient archive and dashboard contexts.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

nmajor25 committed:
Subscribe to ws notifications on patient dashboard & archive

nmajor25 commented on src/js/apps/patients/patient/dashboard/dashboard_app.js:
```diff
@@ -25,21 +25,68 @@ export default App.extend({
-    this.collection = new Backbone.Collection([...actions.models, ...flows.models]);
+    // remove these when BE api endpoint `filter[state]=` is fixed
+    const doneActions = actions.filter(action => !action.isDone());
+    const doneFlows = flows.filter(flow => !flow.isDone());
```
Without this, we'll need to rethink how we determine if a model needs to be fetched or not.

We currently do a `if (this.collection.get(model))`. But without this filtering ☝️, that check won't work in all scenarios.

paulfalgout replied:
Why? If it is already in the collection you should never need to fetch it because you are maintaining the data with notifications even if it is not visible

nmajor25 replied:
Ya I agree, don't think it's needed.

nmajor25 commented on src/js/apps/patients/patient/archive/archive_app.js:
```diff
@@ -13,15 +13,63 @@ export default App.extend({
+    });
+
+    this.listenTo(channel, 'message', ({ category, resource, payload }) => {
+      const modelResource = category === 'ResourceCreated' ? payload.resource : resource;
```
This is technically possible? For a flow or action to be created with a `Done` state? Does that ever happen in the wild? But we should probably handle it either way, right?

paulfalgout replied:
Well we can probably avoid it for now because it probably doesn't happen and we won't need this after the BE change to `ResourceCreated`.   Doing this now just mean we have to test for it.

nmajor25 commented on src/js/views/patients/patient/dashboard/dashboard_views.js:
```diff
@@ -200,7 +200,7 @@ const FlowItemView = View.extend({
     'change:_owner': 'render',
   },
-  behaviors: [RowBehavior],
+  behaviors: [RowBehavior, DoneBehavior],
```
`DoneBehavior` is when the `model.state` changes to `isDone === true`. The action/flow is removed from the list with a fade-out animation.

This was never added for flow list items because it wasn't possible for that to change before. The flow state button is disabled. But now it can change via web socket notification.

paulfalgout replied:
this is nice how easy that is to implement

nmajor25 commented on src/js/apps/patients/patient/dashboard/dashboard_app.js:
```diff
@@ -25,21 +25,68 @@ export default App.extend({
+        return;
+      }
+
+      this._fetchAction(model);
```
Maybe we can consolidate this to one `this._fetchModel(model)` via `Radio.request('entities', 'fetch:${ type }:model', model.id);`...

paulfalgout commented on src/js/apps/patients/patient/archive/archive_app.js:
```diff
@@ -13,15 +13,63 @@ export default App.extend({
   onStart({ patient }, actions, flows) {
-    this.collection = new Backbone.Collection([...actions.models, ...flows.models]);
+    // remove these when BE api endpoint `filter[state]=` is fixed
+    const doneActions = actions.filter(action => action.isDone());
```
I don't think we need to do this, the view is handling the filtering enough

paulfalgout commented on src/js/apps/patients/patient/archive/archive_app.js:
```diff
@@ -13,15 +13,63 @@ export default App.extend({
+    const channel = Radio.channel('ws');
+
+    channel.request('subscribe', this.collection.models, {
+      filters: { actions: this.filter, flows: this.filter },
```
I think both of these filters must include `patientId`

nmajor25 replied:
That seems right. 

But that will subscribe us to each flow's child actions. So when we receive a ws message for one of those, how will we know if it's an action that should be fetched or not? We'd have to check if that action belonged to a flow, but we won't know that until after we fetch it, right?

Or maybe I'm overthinking it? I need to investigate it more next week.

nmajor25 replied:
I think we'll have to check if the action has a parent flow after fetching it. And if it does, don't add it to the collection.

 I don't think we can determine that until after the action has been fetched.

```js
model.fetch().then(() => {
  if (model.type === 'patient-actions' && model.getFlow()) return;

  this.collection.add(model);

  Radio.request('ws', 'add', model);
});
```

paulfalgout replied:
@DrechslerDerek we need to have a way to indicate we want to subscribe to actions for a patient that _aren't_ related to a flow.. ideas?

shadowhand replied:
@paulfalgout I wonder if we need to have two different `types` for patient/flow actions? If we went that route, we might have `patient-actions` and `flow-actions` and then `actions` is a union of them?

shadowhand replied:
Actually, I think it should be `actions-of-flow` and `actions-of-patient`. Then sorting places them together so there is a visual relationship.

paulfalgout replied:
Hmm that's a pretty big change and there's been some discussion about decoupling actions from patients at some point, so I'd personally want to avoid it.  Really I wish the current type was just `actions`   I'm wondering if we can specify `filters.actions.flow = null`, or maybe `filters.patientActions` instead of `filters.actions`, to indicate we want patient actions that aren't associated with a flow..  then if there's some reason to split the types we can do so later?

shadowhand replied:
@paulfalgout if that's not low hanging fruit, then shelve it for now. We can certainly make the current type `actions` with minimal effort, it just requires coordination, right?

In any case... regarding the filter. 😅 I don't like `filter.actions.flow = null` (because PHP can't differentiate between `null` and `undefined` like JS can), but I think `filters.actions.only = 'patient'` (implied enum `patient|flow`) would work?

shadowhand replied:
But I am also open to `filters.patientActions`. Let's see what makes more sense to @DrechslerDerek.

paulfalgout replied:
Worth mentioning we'd never use `only = 'flow'` as for that we just set flow to a particular id.  this is just an issue of filtering out flow actions for a patient filter

DrechslerDerek replied:
I think there could be 2 easy fixes, the first one would be having a flag that says, no flow or flow = null or something, that would be an extra scan condition for dynamo db that would be added dynamically if this flag was set. The other option I see is to do some trick with the current flow filter, where you explicitly pass like 
```
{
    filters: {
        actions: {
            flows: 'null' <or maybe null (at some point in BE this will be a string anyway)>
        }
    }
}
```

which you could actually do right now and it would be a one line change to fix the `$action->flowId ??`. I might be missing how this would cause other issues, and it is more of a bandaid/nuanced solution, but we could do it easily.

shadowhand replied:
If we're going to do some shenanigans like that, it should use the constant value UUID::NIL.

DrechslerDerek replied:
Now I don't think this will work, we need to add another DDB `and` or another filter, meaning the flow on the action is dynamic.

paulfalgout commented on src/js/apps/patients/patient/archive/archive_app.js:
```diff
@@ -13,15 +13,63 @@ export default App.extend({
+
+      if (this.collection.get(model) || model.isFetching()) return;
+
+      if (modelResource.type === 'flows') {
```
yeah I'm perfectly fine combining these with `fetch:${ model.type }:model`   we can probably recombine these methods now since they're simpler than the worklist

paulfalgout replied:
in fact..  I think we will _always_ have the context needed that this can be
```js
const model = Radio.request('entities', 'get:store', modelResource);

if (this.collection.get(model) || model.isFetching()) return;
model.fetch().then(() => { 
  this.collection.add(model);

  Radio.request('ws', 'add', model);
});
```

unless I'm missing something?

nmajor25 replied:
Yup that works 👍

nmajor25 committed:
Add ws cypress tests for dashboard

nmajor25 committed:
Add ws cypress tests for archive

nmajor25 committed:
Add `flow: NIL_UUID` to actions filters

So we don't receive ws notifications for actions that have a parent flow

nmajor25 committed:
Add `cy.wait(200)` to tests

The `mockWs` command is sometimes not finishing before the first `cy.sendWs()` is sent. Which results in the test failing.

This gives `mockWs` more time to initiate the first ws connection.

nmajor25 commented on test/integration/patients/patient/archive.js:
```diff
@@ -652,7 +655,10 @@ context('patient archive page', function() {
+      .wait('@routePatientActions')
+      .wait('@routePatientFlows');
+
+    cy.wait(200);
```
Pushing this up to see if it fixes things when cypress runs in circleci. 

Something is running out of sync. I think it's receiving the first ws notification before the subscription is created. So the app isn't ready to handle it yet. But only sometimes. When I run them locally, it's 50/50 if the test passes or not.

Is there something we're not `.wait()`ing for properly? I can't tell yet.

nmajor25 commented on test/integration/patients/patient/archive.js:
```diff
@@ -463,7 +463,9 @@ context('patient archive page', function() {
-      .wait('@routePatientFlows');
+      .wait('@routePatientActions')
+      .wait('@routePatientFlows')
+      .wait(200); // for initial ws connection to be created
```
Without this `cy.wait()`, these tests fail ~50% of the time. I think what's happening is the `mockWs()` command isn't always finished before we send the first ws message in the test via `cy.sendWs()`.

Adding a `cy.wait(200)` seems to fix the issue, where the test passes 100% of the time. I don't think this does much harm. 

But I would prefer to not do this if there was another way to fix it. The `mockWs` would have to be reconfigured somehow, to somehow let us know when the initial socket connection is ready.

paulfalgout replied:
```js
    cy
      .get('@wsHandleMessage')
      .should('have.been.calledOnce');
```
will do it.

The issue was that the `send` is happening in Cypress before the `subscribe` happens from the app.. so we have to wait until the subscribe is called.

paulfalgout commented on src/js/apps/patients/patient/archive/archive_app.js:
```diff
@@ -1,19 +1,22 @@
     const currentWorkspace = Radio.request('workspace', 'current');
-    const states = currentWorkspace.getStates();
-    const filter = { states: states.groupByDone().done.getFilterIds() };
+    this.states = currentWorkspace.getStates();
```
I think it'd be better to set this in the `onBeforeStart` and not attach things in the `beforeStart`.. and you can probably save the `states` since it's reused in the `subscribe`

nmajor25 replied:
Like this: https://github.com/RoundingWell/care-ops-frontend/pull/1400/commits/af63bdd3d766db9857e159ffa369e8e08fe22252?

nmajor25 committed:
Refactor `this.states` in archive & pt dashboard apps

paulfalgout commented:
@nmajor25 I think this could be rebased and modified with the new service request and maybe it'd work?

nmajor25 commented:
@paulfalgout think so, I'll give it a try 👍

nmajor25 committed:
Refactor subscribe() to use new ws service request

nmajor25 commented on src/js/apps/patients/patient/archive/archive_app.js:
```diff
@@ -22,6 +27,21 @@ export default App.extend({
+      filters: { actions: { ...filters, flow: NIL_UUID }, flows: filters },
+    });
+    Radio.request('ws', 'manage:add', this, this.collection, 'flows');
+    Radio.request('ws', 'manage:add', this, this.collection, 'patient-actions');
```
Two separate service requests here, one for flows and another for actions.

nmajor25 replied:
I tested this locally for both patient dashboard and archive, it worked well.

paulfalgout committed:
Replace test waits with stub calledOnce
